modules/functional_expansion_tools/src/functions/MemoizedFunctionInterface.C

```diff
@@ -0,0 +1,85 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MemoizedFunctionInterface.h"
+
+template <>
+InputParameters
+validParams<MemoizedFunctionInterface>()
+{
+  InputParameters params = validParams<Function>();
+
+  params.addClassDescription("The function uses a cache to potentially reduce the computational "
+                             "burden of reusing a complex or costly function");
+
+  params.addParam<bool>("enable_cache",
+                        false,
+                        "Enables cached function evaluations. Recommended only if this function is "
+                        "used directly in a BC or Kernel. This will be enabled automatically if "
+                        "any of the FX-based BCs are used.");
+
+  params.addParam<bool>("respect_time", false, "Enable to clear the cache at each new time step.");
+
+  return params;
+}
+
+MemoizedFunctionInterface::MemoizedFunctionInterface(const InputParameters & parameters)
+  : Function(parameters),
+    _enable_cache(getParam<bool>("enable_cache")),
+    _respect_time(getParam<bool>("respect_time"))
+{
+}
+
+void
+MemoizedFunctionInterface::meshChanged()
+{
+  // The mesh has changed, which invalidates the cache
+  invalidateCache();
+}
+
+Real
+MemoizedFunctionInterface::value(Real time, const Point & point)
```

Comment:
This is not threadsafe.  You need to use a "doubly checked lock" with a C++11 memory fence.

The one called: "Using C++11 Acquire and Release Fences" here: http://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

Reply:
Would it be possible for multiple threads to be evaluating the same `Point`? If so, I will probably need to lock the entire function body. That may incur a measurable performance hit.

Otherwise, I'm thinking it will be enough to lock just the writing operations.